transformations/writing_system_replacement/transformation.py

```diff
@@ -0,0 +1,519 @@
+    >>> new_text,system_name,_,_ = replace_writing_system("I love potatoes", test_rnd_generator); new_text, system_name
+    ('驿掩㑇㕶誨', 'logographic')
+    >>> new_text,system_name,_,_ = replace_writing_system("I love potatoes", test_rnd_generator); new_text, system_name
+    ('之 笓䒉㘔䆇 躓䒉蝲討蝲䒉䆇䁣', 'alphabet')
+    """
+    writing_systems = {
+        "alphabet": to_alphabet,
+        "syllabary": to_syllabary,
+        "partial_phonemic": to_partial_phonemic,
+        "logographic": to_logographic,
+    }
+    ind = int(rnd_generator.integers(low=0, high=len(writing_systems), size=1))
+    system_name = list(writing_systems)[ind]
+
+    new_text, charset, replacement_report = writing_systems[system_name](
+        input_text, rnd_generator
+    )
+    return new_text, system_name, charset, replacement_report
+
+
+def add_common_writing_system_augments(input_text, rnd_generator):
+    """Randomly selects 0 or several augmentations, and applies them the the input text.
+
+    For example, it could reverse the writing direction.
+
+    >>> test_rnd_generator = np.random.default_rng(seed=42)
+    >>> add_common_writing_system_augments("I love carrots, and potatoes!", test_rnd_generator)
+    ('Ilovecarrotsandpotatoes', ['remove_punctuation', 'remove_spaces'])
+    >>> add_common_writing_system_augments("I love carrots, and potatoes!", test_rnd_generator)
+    ('I love carrots an dpotatoe s', ['remove_punctuation', 'rearrange_spaces'])
+    >>> add_common_writing_system_augments("I love carrots, and potatoes!", test_rnd_generator)
+    ('Ilovecarrots,andpotatoes!', ['remove_spaces'])
+    >>> add_common_writing_system_augments("I love carrots, and potatoes!", test_rnd_generator)
+    ('seota t op dnastorra ce v o lI', ['reverse_writing_direction', 'remove_punctuation', 'remove_spaces', 'rearrange_spaces'])
+    """
+    augments = {
+        "reverse_writing_direction": reverse_writing_direction,
+        "remove_punctuation": remove_punctuation,
+        "remove_spaces": remove_spaces,
+        "rearrange_spaces": rearrange_spaces,
+    }
+
+    indexes = rnd_generator.integers(
+        low=0, high=2, size=len(augments)
+    ).tolist()
+    augments_to_use = []
+    augments_list = list(augments)
+    for i in range(len(indexes)):
+        ind = indexes[i]
+        if ind == 1:
+            augments_to_use.append(augments_list[i])
+
+    for augment_name in augments_to_use:
+        input_text = augments[augment_name](input_text, rnd_generator)
+
+    return input_text, augments_to_use
+
+
+class WritingSystemReplacement(SentenceOperation):
+    tasks = [
+        TaskType.TEXT_CLASSIFICATION,
+        TaskType.TEXT_TO_TEXT_GENERATION,
+    ]
+    languages = "All"
+    keywords = [
+        "noise",
+        "morphological",
+        "lexical",
+        "syntactic",
+        "rule-based",
+        "unnaturally-written",
+        "possible-meaning-alteration",
+        "high-coverage",
+    ]
+
+    def __init__(self, seed=0, max_outputs=1):
```

Comment:
Is there a reason why many of the function parameters are hardcoded instead of passed as class parameters here and saved as class attributes? For example `replace_white_spaces`, `use_all_codes`, `exclude_white_spaces`, `main_rnd_generator`, `spaces_frequency`

Similarly, is there a reason for `chinese_chars` to be a global variable instead of a class attribute?

Reply:
Good point! I've re-implemented them as class attributes.